### DISCUSSION TITLE
Session-token cookie not set

### DISCUSSION THREAD

**Discussion by Mikhail1Demianenko** [nextauthjs/next-auth#4089] [Help]
### Question 💬

Hello community.
Thank you for an amazing library!

I am trying to create a nextjs app with many subdomains, such that the user can authenticate using the apex domain (no subdomain) and share the auth state across the rest subdomains.
This issue has been discussed [here](https://github.com/nextauthjs/next-auth/issues/405), [here](https://github.com/nextauthjs/next-auth/issues/1668) and documented [here](https://next-auth.js.org/configuration/options#cookies).

Basically, what is suggested is to add the following object into the object we pass to the `NextAuth` function in `[...nextauth].js`:

```  
cookies: {
    sessionToken: {
      name: "next-auth.session-token",
      options: {
        domain: ".localhost",
        path: "/",
        httpOnly: true,
        sameSite: "lax",
        secure: false
      }
    }
  }
```
(In the object I hardcoded the values for local dev env, so it is easier to read)
With the key thing in the object being the dot before the domain.

Without this object, the cookie after login is set correctly:
<img width="946" alt="image" src="https://user-images.githubusercontent.com/69697322/156061869-43d3449d-db21-4317-8dbb-3e18e47fdbae.png">
And the **cookies set in the browser are just 5 in total**

However, **with** that object I have the following picture:

<img width="693" alt="image" src="https://user-images.githubusercontent.com/69697322/156062316-8ba05acc-488b-469f-af49-c53b25ba052e.png">

The _session-token_ cookie has disappeared. And, of course, authentication status results in `unauthenticated`, even though a session gets created.

Moreover, browser now has **25** cookies set:
<img width="329" alt="image" src="https://user-images.githubusercontent.com/69697322/156062812-1b1555f4-3f15-4053-a6fa-b1ec452d7d9d.png">

<img width="440" alt="image" src="https://user-images.githubusercontent.com/69697322/156062838-e5230d41-5fde-455a-ad7c-6ac1d389a3bb.png">
 
Having gone through all the above posts it seems to have worked for others, but sadly not for me

Does anyone have any idea what I am doing wrong?

Thank you!

### How to reproduce ☕️

`[...nextauth].js` file with `next-auth v.4.2.1`:
```
const prisma = new PrismaClient();

export default NextAuth({
  adapter: PrismaAdapter(prisma),
  cookies: {
    sessionToken: {
      name: "next-auth.session-token",
      options: {
        domain: ".localhost",
        path: "/",
        httpOnly: true,
        sameSite: "lax",
        secure: false
      }
    }
  },
  providers: [
    Google({
      clientId: 'clientId',
      clientSecret: 'clientSecret'
    })
  ],
  secret: 'secret'
  session: {
    strategy: "database",
    maxAge: 30 * 24 * 60 * 60,
  },
  pages: {
    signIn: '/auth/login',
  }
})
```

### Contributing 🙌🏽

Yes, I am willing to help answer this question in a PR

**Comment by umerjaved178** [comment]
Hey @Mikhail1Demianenko I am facing the same issue, as soon as I add this object
```
cookies: {
    sessionToken: {
      name: "next-auth.session-token",
      options: {
        domain: ".localhost",
        path: "/",
        httpOnly: true,
        sameSite: "lax",
        secure: false
      }
    }
  }
  ```
  
  Session cookie won't get generated and thus result in an unauthenticated session.
  It is a bug on next-auth side with the new version. I have experimentally installed v3 of next-auth, session is getting authenticated, cookie name is changing to my custom one, but ofcourse 'domain' is not working as it was not supported in that version
 Team can you please look into this issue and moved it from discussion to BUGS

**Reply by Mikhail1Demianenko** [reply]
Hi there,

Thank you for your reply
As the matter of fact, I also tried do set the cookie manually through browser and tried to change `domain` of already existing one. Result - the cookie wouldn't get set or got deleted. The problem therefore was not next-auth, I figured out later, but the `localhost` domain: there is a policy in browsers restraining you from adding a leading dot for localhost (i dont know the details)
So what I did instead is I added the following lines to the `/etc/hosts`:
```
127.0.0.1	test.com
127.0.0.1	app.test.com
```
And the cookie was then available on app. subdomain.
As a heads-up, if you go with this option, you will have to type `test.com:3000` instead of `localhost:3000` in the URL bar and `/etc/hosts` does not allow a wildcard subdomains, so I simply created another record like '127.0.0.1	sub.test.com', which is good enough for me to represent all subdomains for dev env

**Reply by umerjaved178** [reply]
@Mikhail1Demianenko  Thanks Mikhail for sharing the solution.
When I am implementing it
1. The App is up on domain **test.com:3000** (not secure), https://paste.pics/3868fa2e7d6510417a459a5e24d63710
2. it redirects to the page to select the google accounts, on selecting the account
3. this pop-ups (https://paste.pics/7a2fd451fe9c9ce9c8469c023aae4ab4) (and domain automatically changes to localhost:3000 at this step), 
4. if I move forward from here, by clicking the sign-in button shown in the screenshot in point 3,
5. it signed me in but on localhost:3000 https://paste.pics/6568697d4002ffa34ae8485ced303d21

in order to fix this, I have tried to set **redirect** callback function in [...nextauth].js file
```
callbacks: {
    async redirect({ url, baseUrl }) {
      return "http://test.com:3000/";
    },
  },
```

but it is also not working, can you please tell me how did you overcome this issue :))

**Reply by Mikhail1Demianenko** [reply]
I think you forgot to change the redirect url in your google credentials console
Let me know if its not the case

**Reply by umerjaved178** [reply]
Oh yes, I forgot, it was set to http://localhost.com:3000/api/auth/callback/google

Now, I have set it to _http://test.com:3000/api/auth/callback/google_ , as we know this is not secure connection (HTTP), google console is not able to save it and giving this error (https://paste.pics/23d45227f2547c8642e58dcdbbddc179)
and setting it to https like _https:test.com......._ , it will not work neither on local nor with google console (https://paste.pics/a95546a249e15c324a23e3a113155d2c)

 
Side note: 
In case you want to see my /etc/hosts file, it looks like this https://paste.pics/b1fb16e4f96fb30302b793afe9938de9

**Reply by umerjaved178** [reply]
@Mikhail1Demianenko  I am wondering how did you overcome this inevitable error 🤔

**Reply by Mikhail1Demianenko** [reply]
I never had this problem in the first place
Here are my GC settings:
<img width="501" alt="image" src="https://user-images.githubusercontent.com/69697322/156728376-eb1453ba-46df-4d5c-8965-3edf4250d03e.png">
Consider aligning yours with mine
Otherwise, you might have a callback hooked up somewhere, I'd search for 'localhost' in the codebase and replace it with `test.com` everywhere it makes sense to do so

**Reply by umerjaved178** [reply]
@Mikhail1Demianenko  🚀
So nice of you man! you have saved my days ☺️ 😊

It is resolved

A few things to avoid for future readers:
1. I couldn't set 'http' on google console because I have selected a few scopes that need only 'https' connection if the domain is other than localhost
2. it is compulsory to set NEXTAUTH_URL to the new domain, you can't leave it empty else it will redirect you to localhost
3. _domain_ value in cookies -> sessionToken must not be just 'test', it should be '.test.com'
4. As earlier mentioned by Mikhail, the problem is with local host domain (some browser security things), you have to change it to a custom domain in /etc/hosts/ file
5. sometimes 'states' error pops up, switch to incognito mode

**Reply by Mikhail1Demianenko** [reply]
Im glad it helped 😊

**Reply by jackrdye** [reply]
> Hi there,
> 
> Thank you for your reply As the matter of fact, I also tried do set the cookie manually through browser and tried to change `domain` of already existing one. Result - the cookie wouldn't get set or got deleted. The problem therefore was not next-auth, I figured out later, but the `localhost` domain: there is a policy in browsers restraining you from adding a leading dot for localhost (i dont know the details) So what I did instead is I added the following lines to the `/etc/hosts`:
> 
> ```
> 127.0.0.1	test.com
> 127.0.0.1	app.test.com
> ```
> 
> And the cookie was then available on app. subdomain. As a heads-up, if you go with this option, you will have to type `test.com:3000` instead of `localhost:3000` in the URL bar and `/etc/hosts` does not allow a wildcard subdomains, so I simply created another record like '127.0.0.1 sub.test.com', which is good enough for me to represent all subdomains for dev env

You are my savior <3. I feel like the docs could definitely be clearer on this setup. Spent 24 hours + on this ... (this should be included in the docs tbh)

**Reply by MatteoGauthier** [reply]
Hey @Mikhail1Demianenko @umerjaved178 @jackrdye ! Sorry to bother you, many people seem to have had this problem before you (including me), do you have a public repository or guide to solve this problem? Have a nice day!

**Reply by EPanni** [reply]
@Mikhail1Demianenko , sorry but how come adding those lines to /etc/hosts solves this problem if when someone else try to log in to the app ( from his/her local machine ) it will not work since they did not set it up?

**Comment by codingcaffeine** [comment]
mine works like this
````
import NextAuth, { NextAuthOptions } from "next-auth"
import GithubProvider from "next-auth/providers/github"
import { PrismaAdapter } from "@next-auth/prisma-adapter";
import prisma from "@/lib/prisma";

// For more information on each option (and a full list of options) go to
// https://next-auth.js.org/configuration/options
export const authOptions: NextAuthOptions = {
  adapter: PrismaAdapter(prisma),
 
  // https://next-auth.js.org/configuration/providers/oauth
  providers: [
    GithubProvider({
      clientId: process.env.GITHUB_ID as string,
      clientSecret: process.env.GITHUB_SECRET as string,
    }),
  ],
  secret: process.env.SECRET,
  session: {
    strategy: 'database'
  }
}

export default NextAuth(authOptions)

**Comment by smcelhinney** [comment]
I'd just like to add that it took me ages to figure this out - but it's mentioned in the first line of the original question. The NextAuth sessionCookie needs to be created on the *apex* domain, in order to be accessible across subdomains. 

When I originally configured my Next.js project, I took sensible defaults which involved them creating a 308 Redirect from mydomain.dev -> www.mydomain.dev, which meant that I was effectively authing on a subdomain.

I had to remove this redirect (I wasn't using it anyway) and it worked.

**Comment by iiiok** [comment]
Next-Auth Session-token issues have been hindering me for half month.
Some time, unable to getSession after Login, some time unable to logout.
--------------Here is how it solved, update Ur next-auth library AND Ur NEXT library !
    "next": "^13.3.1",
    "next-auth": "^4.22.1",

**Comment by Strajk** [comment]
For people arriving here when googling for `disappearing Session-token cookie`:

Check that all your `getServerSession(...)` calls are passed same config as the third arg, as mentioned in [docs](https://next-auth.js.org/configuration/nextjs)

I had following in my codebase, and the second line caused session-token to be overridden to empty string, thus causing log out
```
getServerSession(ctx.req, ctx.res, authOptions) // in one file
getServerSession(ctx.req, ctx.res, {}) // in another file
```

**Comment by trevorpfiz** [comment]
anyone have success getting the session token cookie in subdomains when using `"next-auth": "5.0.0-beta.4",`? will update here if I figure anything out.

**Reply by trevorpfiz** [reply]
#9631

**Reply by trevorpfiz** [reply]
my current workaround is to not wrap the middleware with auth, I just use it like this:

```
export default async function middleware(req: NextRequest) {
  const url = req.nextUrl;
  const session = await auth();
  const isLoggedIn = !!session?.user;
```

**Reply by atalaykarahan** [reply]
Hi i have the same next-auth version and my project already deployed but whenever i deploy my project my entire cookie has domain error. But if i try in local everything is perfect and i cant find anything about "auth.js cookie domain settings" documents or something like that do you have the same issue
![image](https://github.com/nextauthjs/next-auth/assets/114542681/dddfa993-e3d6-40f1-bb22-2a64c3b3d458)
how can i solved this issue can you help me?
